### DISCUSSION TITLE
How to control when timeout occurs?

### DISCUSSION THREAD

**Discussion by andybrandt** [modelcontextprotocol/modelcontextprotocol#1207] [Q&A - Server implementation]
### Pre-submission Checklist

- [X] I have checked that this question would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions and documentation for answers

### Question Category

- [ ] Protocol Specification
- [X] SDK Usage
- [X] Server Implementation
- [ ] General Implementation
- [ ] Documentation
- [ ] Other

### Your Question

Using Python SDK how can I set how long MCP implementation should wait for the response from an external resource? I am building an mcp server which talks to a resource that sometimes takes about 20-40 seconds to respond, so the response is never returned because a timeout occurs. Looked at the docs but I don't see indication how to do this.

**Comment by jspahrsummers** [comment]
Timeouts are controlled on the client side, not the server side. This is necessary because a misbehaving server may _never_ respond (even if it "meant to").

You may want to consider prefetching the resource, or introducing some kind of multi-step loading, like:
1. Client sends a request. Loading begins, server responds with "in progress."
3. Client polls until loading is complete and/or server sends a notification that loading completed.
4. Client sends a final request, and gets back the actual resource.

Though this question did make me think that we could support [auto-extending timeouts when progress notifications are used](https://github.com/modelcontextprotocol/typescript-sdk/issues/87), which doesn't currently happen.

**Comment by andybrandt** [comment]
Thanks tons for the response about timeouts! 

Reading the MCP Python SDK documentation, I see examples of progress tracking using request_context.meta.progressToken, but when I tried this it turned out these objects (context and progressToken) are None when accessed through server.request_context. 

Is this because progress notification support needs to be implemented on the client side first? The documentation suggests this functionality but if I understand correctly, I should focus on implementing the multi-step approach you suggested instead of trying to use progress notifications at this point?

**Reply by jspahrsummers** [reply]
Hmm, progress notifications from the server to the client should work. Perhaps @dsp-ant knows more here 🙏

**Reply by andybrandt** [reply]
Well, they do not allow to keep the session alive - the timeout still occurs. 
So for now I solved it by sending back to using AI information that the response is not yet ready, Claude is smart enough to wait a bit and try again.

**Comment by friendlygeorge** [comment]
Adding to the suggestions above — the timeout is indeed client-side, but there are practical server-side patterns I have used in the system-monitoring and cron-scheduler servers (`@supernova123/system-monitoring-mcp-server`, `@supernova123/cron-scheduler-mcp-server`) when a tool legitimately needs more than 30 seconds:

**Pattern: Two-phase tool call**

```typescript
// Tool returns immediately with a job ID, then a separate tool checks status
server.tool('run-long-task', async (args) => {
  const jobId = await scheduler.enqueue(args);
  return { content: [{ type: 'text', text: JSON.stringify({ jobId, status: 'queued' }) }] };
});

server.tool('get-task-status', async (args) => {
  const result = await scheduler.getResult(args.jobId);
  return { content: [{ type: 'text', text: JSON.stringify(result) }] };
});
```

**What has worked in production:**
- Stream progress via `meta.progressToken` — works today in Claude Desktop
- Pre-flight cache layer in front of the slow upstream
- For truly batch jobs, expose them as scheduled tasks rather than request/response

The cron-scheduler server is built around exactly this pattern — scheduling work that runs independently of the request/response cycle.

The key insight: if a tool takes >30s, it probably should not be a synchronous tool call. Reframe it as a queued job with a status check, and the timeout problem disappears.

**Comment by Erfouni** [comment]
The original answer is still correct about one important boundary: **the server cannot extend a timeout chosen by the client**. However, MCP 2025-11-25 added an experimental protocol-native option for this exact class of long-running work: **Tasks**.

If both sides support it, the flow is now:

1. The server advertises `capabilities.tasks.requests.tools.call`.
2. The tool advertises `execution.taskSupport` as `optional` or `required`.
3. The client sends a task-augmented `tools/call`.
4. The server quickly returns a `CreateTaskResult` with `taskId`, status, TTL, and polling interval.
5. The client uses `tasks/get` for status and `tasks/result` for the eventual `CallToolResult`.

This prevents the 20–40 second upstream call from having to fit inside one synchronous request timeout, while keeping cancellation, deferred result retrieval, and reconnection semantics in the protocol.

A couple of practical details:

- A missing `progressToken` is expected when the client did not include one in the request `_meta`. Progress is opt-in and a server should not assume every client requests or supports it.
- Progress notifications are useful UX, but they do **not** universally guarantee that a host will extend its request deadline. Do not use them as the only timeout strategy.
- Keep a separate timeout around the external HTTP/resource call inside the server. The client timeout and the upstream timeout protect different boundaries.
- Tasks are still experimental and host support varies. If your target client does not support them, the two-tool fallback remains valid: `start_job(...) -> jobId`, then `get_job(jobId)` (and optionally `cancel_job(jobId)`). Return the first response immediately and store job state outside the request handler.

So the implementation choice is:

- **Task-capable client:** use MCP Tasks.
- **Legacy/non-task client:** use the explicit queued-job/status-tool pattern.
- **Short, reliably bounded upstream call:** a normal synchronous tool call is still fine; configure the *client* deadline there.

Official references:

- [MCP Tasks (2025-11-25)](https://modelcontextprotocol.io/specification/2025-11-25/basic/utilities/tasks)
- [Task extension overview and client support](https://modelcontextprotocol.io/extensions/tasks/overview)
- [Progress notifications](https://modelcontextprotocol.io/specification/2025-11-25/basic/utilities/progress)

**Comment by localden** [comment]
We've limited Discussions in this repository to meeting notes and are closing this thread. To continue the conversation, pick a channel from [Contributor Communication](https://modelcontextprotocol.io/community/communication). Thanks for posting.
